### DISCUSSION TITLE
git-lfs support

### DISCUSSION THREAD

**Discussion by extrawurst** [extrawurst/gitui#1089] [Ideas]
I was recently asked if git-lfs is supported. 
Research shows that it is not supported out of the box in our upstream `libgit2`:
* [git-lfs#375](https://github.com/git-lfs/git-lfs/issues/375)
* [nodegit#914](https://github.com/nodegit/nodegit/issues/914)
* [libgitsharp2#1236](https://github.com/libgit2/libgit2sharp/issues/1236)

No idea how git-lfs is implemented in itself but it seems to be a manual job to be supported in gitui.

_edit:_
seems lfs is using 4 git hooks to work (none of which are currently supported by gitui):
* post-checkout
* post-commit
* post-merge
* pre-push (✅ as of 0.28)

maybe this can be used as a reference implementation:
https://github.com/Axosoft/nodegit-lfs

**Comment by extrawurst** [comment]
the libgit2 maintainers keep referring to the filters to implement git-lfs, in order to do this in rust git2-rs needs to support the upstream functions: https://github.com/rust-lang/git2-rs/issues/442

**Reply by extrawurst** [reply]
here is a another helpful comment on this: https://github.com/libgit2/libgit2/issues/6565#issuecomment-1551331883

**Comment by strogonoff** [comment]
There’s a tiny LFS compatibility layer I wrote for isomorphic-git: https://github.com/riboseinc/isogit-lfs. 

I found it reasonably straightforward to do the following bare minimum needed for our use case:

1) Given a blob, perform an appropriate upload request to LFS server, and get LFS pointer metadata as a response.

2) Serialize received LFS pointer metadata and store it at object’s path in the repository (in place of the original blob).

3) Parse/deserialize LFS pointer metadata stored at an object’s path.

4) Given LFS pointer metadata, perform an appropriate download request to LFS server, and receive a blob as response.

5) Write received blob to a correct LFS cache path in Git directory.

In case of gitui, users would probably expect to be able to check out LFS blobs into working directory, but it wasn’t required in our use case.

**Reply by extrawurst** [reply]
Ok I only glanced over it right now.
> Works in Node runtime only
that is kind of a problem to impose as a dependency for anyone using gitui.

Also I have no idea how to integrate this. Fancy giving me a rough summary?

**Comment by austinbutler** [comment]
Is this the same thing blocking https://github.com/extrawurst/gitui/issues/1468? I found that `gitui` doesn't run the smudge filter on staging a file, while the Git CLI, Neogit, and Lazygit all do.

**Comment by extrawurst** [comment]
Now properly tracked in #2812
